adapters/metax/metax.go

```diff
@@ -0,0 +1,244 @@
+	if err := json.Unmarshal(respData.Body, &bidResp); err != nil {
+		return nil, []error{err}
+	}
+
+	// additional no content check
+	if len(bidResp.SeatBid) == 0 || len(bidResp.SeatBid[0].Bid) == 0 {
+		return nil, nil
+	}
+
+	resp := adapters.NewBidderResponseWithBidsCapacity(len(bidResp.SeatBid[0].Bid))
+	for _, sb := range bidResp.SeatBid {
+		for i := range sb.Bid {
+			bid := &sb.Bid[i]
+			resp.Bids = append(resp.Bids, &adapters.TypedBid{
+				Bid:     bid,
+				BidType: getBidType(bidReq.Imp, bid.ImpID),
+			})
+		}
+	}
+	return resp, nil
+}
+
+func (a *adapter) getEndpoint(ext *openrtb_ext.ExtImpMetaX) (string, error) {
+	params := macros.EndpointTemplateParams{
+		PublisherID: url.PathEscape(ext.PublisherID),
+		AdUnit:      url.PathEscape(ext.Adunit),
+	}
+	return macros.ResolveMacros(a.template, params)
+}
+
+func parseBidderExt(imp *openrtb2.Imp) (*openrtb_ext.ExtImpMetaX, error) {
+	var bidderExt adapters.ExtImpBidder
+	if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+		return nil, err
+	}
+
+	var metaxExt openrtb_ext.ExtImpMetaX
+	if err := json.Unmarshal(bidderExt.Bidder, &metaxExt); err != nil {
+		return nil, errors.New("Wrong MetaX bidder ext")
+	}
+
+	return &metaxExt, nil
+}
+
+// publisher ID and adunit is numeric at this moment
+func validateParams(ext *openrtb_ext.ExtImpMetaX) error {
+	if _, err := strconv.Atoi(ext.PublisherID); err != nil {
+		return errors.New("invalid publisher ID")
+	}
+	if _, err := strconv.Atoi(ext.Adunit); err != nil {
+		return errors.New("invalid adunit")
+	}
+	return nil
+}
+
+func preprocessImp(imp *openrtb2.Imp) error {
+	if imp == nil {
+		return errors.New("imp is nil")
+	}
+
+	if imp.Banner != nil {
+		bannerCopy, err := assignBannerSize(imp.Banner)
+		if err != nil {
+			return err
+		}
+		imp.Banner = bannerCopy
+	}
+
+	// clean inventory
+	switch {
+	case imp.Video != nil:
+		imp.Banner = nil
+		imp.Native = nil
+		imp.Audio = nil
+	case imp.Banner != nil:
+		imp.Video = nil
+		imp.Native = nil
+		imp.Audio = nil
+	default:
+	}
+
+	return nil
+}
+
+func assignBannerSize(banner *openrtb2.Banner) (*openrtb2.Banner, error) {
+	if banner.W != nil && banner.H != nil {
+		return banner, nil
+	}
+
+	if len(banner.Format) == 0 {
+		return banner, nil
+	}
+
+	return assignBannerWidthAndHeight(banner, banner.Format[0].W, banner.Format[0].H), nil
+}
+
+func assignBannerWidthAndHeight(banner *openrtb2.Banner, w, h int64) *openrtb2.Banner {
+	bannerCopy := *banner
+	bannerCopy.W = ptrutil.ToPtr(w)
+	bannerCopy.H = ptrutil.ToPtr(h)
+	return &bannerCopy
+}
+
+func getBidType(imps []openrtb2.Imp, impID string) openrtb_ext.BidType {
+	for _, imp := range imps {
+		if imp.ID != impID {
+			continue
+		}
+		switch {
+		case imp.Banner != nil:
+			return openrtb_ext.BidTypeBanner
+		case imp.Video != nil:
+			return openrtb_ext.BidTypeVideo
+		case imp.Native != nil:
+			return openrtb_ext.BidTypeNative
+		case imp.Audio != nil:
+			return openrtb_ext.BidTypeAudio
+		default:
```

Comment:
As per `static/bidder-info/metax.yaml`, only `banner video` are supported media types.

Therefore should update adapter code to remove support for `native audio` media types. Alternatively, could update bidder info yaml to support  `native audio` media types.